Universal builds for macOS

simonjwright opened the pull request:
For macOS (Darwin), some users are running on Intel silicon (`x86_64`), some on Apple silicon (`aarch64`). The purpose of this PR is to generate universal binaries of `alr` for macOS, so that users won't need to specify which version they need -- the loader will select the appropriate binary.

To do this, in both the `nightly.yml` and the `ci-macos.yml` workflows  a new job `build-macos-universal` runs after the `build` job has completed (which implies completion of its runs on `macos-13` (`x86_64`) and `macos-14` or `macos-latest` (`aarch64`)).

This job dowloads & unpacks the `x86_64` build to `x86_64/`, and the `aarch64` build to `aarch64/`. It then runs `lipo` to create the universal binary `alr`. The new binary has to be marked as executable, don't know why.

`alr-<release>-bin-universal-macos.zip` is created and uploaded to the release directory.

Notes:
* All three binaries (`aarch64`, `x86_64` and `universal` are created, mainly because I didn’t know how otherwise to preserve the outputs of the two parallel build matrix runs. This could cause confusion, which should be minimised when we update the ada-lang-io and alire.ada.dev sites
* I didn’t know how to get the metadata for a build (`RELEASE.dat` etc), so I copied it from the `aarch64` build.

simonjwright committed:
Nightly build generates macOS universal binary.

For macOS (Darwin), some users are running on Intel silicon (x86_64),
some on Apple silicon (aarch64). The purpose of this commit is to
generate a universal binary of alr for macOS, so that users won't need
to specify which version they need -- the loader will select the
appropriate binary.

To do this, a new job 'build-macos-universal' runs after the 'build'
job has completed (which implies completion of its runs on macos-13
(x86_64) and macos-14 (aarch64)).

This job dowloads & unpacks the x86_64 build to x86_64/, and the
aarch64 build to aarch64/. It then copies aarch64/ to universal/ (so
as to include the additional information, and runs lipo to create the
universal binary in universal/bin/alr. The new binary has to be marked
as executable, don't know why.

alr-nightly-bin-universal-macos.zip is created and uploaded to the
release directory.

  * .github/workflows/nightly.yml (build-macos-universal): new.

simonjwright committed:
Release build generates macOS universal binary.

For macOS (Darwin), some users are running on Intel silicon (x86_64),
some on Apple silicon (aarch64). The purpose of this commit is to
generate a universal binary of alr for macOS, so that users won't need
to specify which version they need -- the loader will select the
appropriate binary.

To do this, a new job 'build-macos-universal' runs after the 'build'
job has completed (which implies completion of its runs on macos-13
(x86_64) and macos-latest (aarch64)).

This job dowloads & unpacks the latest prerelease x86_64 build to
x86_64/, and the latest prerelease aarch64 build to aarch64/. It then
runs lipo to create the universal binary in ./bin/alr. The new binary
has to be marked as executable, don't know why.

alr-<release]-bin-universal-macos.zip is created and uploaded to the
release directory.

  * .github/workflows/ci-macos.yml (build-macos-universal): new.

simonjwright committed:
Merge branch 'master' into universal-build

simonjwright commented:
Updated from master branch.

simonjwright commented:
> The purpose of this PR is to generate universal binaries of `alr` for macOS, so that users won't need to specify which version they need -- the loader will select the appropriate binary.

Badly phrased, I think - "the loader will select the appropriate architecture from the single binary executable" might be clearer

mosteo commented on .github/workflows/nightly.yml:
```diff
@@ -113,3 +114,53 @@ jobs:
+          out-file-path: 'x86_64'
+          extract: true
+      - name: check x86_64 attributes
+        run: xattr -vr x86_64
```
Is this a debug leftover? Or missing in `ci-macos`?

simonjwright replied:
Yes, it's a debug thing -- well, really investigative; since the download is via a url, I was worried that the `com.apple.quarantine` attribute might get applied to the download contents, which would mean that when a user downloads the .zip via curl they would still have to remove the attribute even though curl doesn't apply it itself. Result of investigation was that there's no quarantine attribute; will remove this step.

mosteo commented on .github/workflows/nightly.yml:
```diff
@@ -113,3 +114,53 @@ jobs:
+          out-file-path: 'aarch64'
+          extract: true
+      - name: check aarch64 attributes
+        run: xattr -vr aarch64
```
Likewise

simonjwright replied:
See above: will remove.

mosteo commented on .github/workflows/ci-macos.yml:
```diff
@@ -113,3 +113,67 @@ jobs:
+          lipo x86_64/bin/alr aarch64/bin/alr -create -output bin/alr
+          chmod +x bin/alr
+          cp aarch64/LICENSE.txt .
+          zip alr-bin-macos.zip bin/alr LICENSE.txt
```
No `-universal-` in the name like in the nightly one?

simonjwright replied:
The [upload step](https://github.com/simonjwright/alire/blob/e6b7401f1c6e70f69a7b5284f7b915a37e7ceb93/.github/workflows/ci-macos.yml#L169-L178), copied from :150 above, uses `alr-bin-macos.zip ` at line 176 and does the renaming.

simonjwright replied:
Resolved in commit 9fbdb2e.

mosteo replied:
Thanks, I missed the renaming in the upload step.

simonjwright committed:
Remove superseded investigative steps.

Remove two steps that were inserted to check that no
com.apple.quarantine attributes were applied when the individual
architecture builds were downloaded (none were applied).

  * .github/workflows/nightly.yml (build-macos-universal):
      (check x86_64 attributes): remove.
      (check aarch64 attributes): remove.
